### ISSUE TITLE
docker_registry_image resource fails creation at getImageDigest even though image was uploaded when using ECR Public registry 

### ISSUE THREAD

**Issue by RutledgePaulV** [kreuzwerker/terraform-provider-docker#356]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and docker Provider) Version

Terraform v1.1.2 on darwin_amd64
provider registry.terraform.io/kreuzwerker/docker v2.16.0

### Affected Resource(s)

* `docker_registry_image`

### Terraform Configuration Files

```hcl
variable public_token {
    type = string
}

variable public_registry_alias {
    type = string
}

provider "docker" {
  registry_auth {
    address  = "public.ecr.aws"
    username = "AWS"
    password = var.public_token
  }
}

resource "docker_registry_image" "this" {
  name          = "public.ecr.aws/${var.public_registry_alias}/my-image:latest"
  keep_remotely = true
  build {
    context    = "${path.root}/dockerfiles/my-image"
  }
}
```

### Debug Output

```
2022-03-06T12:07:23.309-0600 [ERROR] vertex "docker_registry_image.this" error: Unable to create image, image not found: unable to get digest: Got bad response from registry: 400 Bad Request
```

### Expected Behaviour

docker_registry_image resource should have been marked as a successful creation since the image was uploaded.

### Actual Behaviour

An error was thrown when fetching the sha to fill in the computed attribute after the image upload: https://github.com/kreuzwerker/terraform-provider-docker/blob/a8517ea3024f681baa609e930120482d09900733/internal/provider/resource_docker_registry_image_funcs.go#L55

### Steps to Reproduce

1. `terraform apply`

### References

* https://docs.aws.amazon.com/AmazonECR/latest/public/public-registries.html#registry_auth_http

**Comment by RutledgePaulV** [comment]
It appears that this provider assumes the registry auth will always be basic auth, but ECR Public actually wants the password (only) sent as a Bearer token when fetching the manifest.

**Comment by RutledgePaulV** [comment]
Unsuccessful:

```bash
curl -v -H "Authorization: Basic $(echo "AWS:$TOKEN" | base64)" "https://public.ecr.aws/v2/<alias>/<image-name>/manifests/<tag>"
```


Successful:

```bash
curl -v -H "Authorization: Bearer $TOKEN" "https://public.ecr.aws/v2/<alias>/<image-name>/manifests/<tag>"
```

**Comment by RutledgePaulV** [comment]
Perhaps we could make a setting available in the auth_config on providers and docker_registry_image resources? 

Something like:

```
manifest_auth_scheme = "bearer" | "basic"
```

**Comment by RutledgePaulV** [comment]
ECR Public expects a bearer token and not basic auth for this endpoint. I noticed the code makes an attempt to handle registries that wan't "oauth" if a 401 is returned, but ECR returns a 400 so it skips the handling completely. The token to send in the manifest request is actually decipherable from the password (base64 decode the password + split on `:` and take the latter half), but that's probably quite specific to ECR.

**Comment by RutledgePaulV** [comment]
A couple questions:

It looks like this is code that bypasses the official docker client and makes http calls to the registry and doesn't properly support all the different authentication mechanisms that remote registries are using.  Is there a reason why the digest needs to be obtained from the remote registry instead of a local image inspection? It seems to me if we can defer to the docker client as much as possible it would free this library from needing to support all the nuances.

**Comment by Junkern** [comment]
Thanks for the submitting the issue and sorry for the long delay.

First question: How are you getting your `var.public_token` value? Via `aws ecr-public get-login-password` or `aws ecr-public get-authorization-token`?

The fix for the ECR `getImageDigestProblem` woud be to add `public.ecr.aws` to https://github.com/kreuzwerker/terraform-provider-docker/blob/master/internal/provider/data_source_docker_registry_image.go#L105
But, in addition to that, the authentication only works if you `base64 -d` the output from the `ecr-public get-authorization-token`, because only then the request is in the correct format....

So, I agree with you that it would be nice to defer as much as possible to the docker client.
However, there is no docker client function to get digest of the image without downloading the whole image. In order to get just the digest, we have to work with the REST endpoint of the registry.

We could, however, defer the whole auth handling to the docker client. I have not found any easy to use API for that, yet, I probably have to keep looking.

In short: Your initial problem can be fixed :)

**Comment by Junkern** [comment]
Improved auth with ECR registries with `v2.19.0`

**Comment by wSedlacek** [comment]
I am running into the same issue with Azure Container Registry when using a token as a password.
https://docs.microsoft.com/en-us/azure/container-registry/container-registry-repository-scoped-permissions

It pushes up the image and it is in the registry, but it fails to get the digest after it pushes.
```
Error: Unable to create image, image not found: unable to get digest: Bad credentials: 401 Unauthorized
```

From what I can tell in making the request in postman, I am getting a `error="malformed_token"` in the `Www-Authenticate`.
I have tried with both the raw token and the token being base64 encoded.

**Comment by wSedlacek** [comment]
Looks like this needs to be used for oauth
https://docs.microsoft.com/en-us/rest/api/containerregistry/access-tokens/get-from-login?tabs=HTTP

**Comment by Junkern** [comment]
@wSedlacek could you file a new issue for that? That helps me to keep better track :)

**Comment by wSedlacek** [comment]
I made a PR here to fix the issue.
https://github.com/kreuzwerker/terraform-provider-docker/pull/451
